File: clients/create/templates/android-minsdk24/feature/users/src/main/kotlin/io/github/ydonghao/{{appPackageSegment}}/feature/users/UsersViewModel.kt
package io.github.ydonghao.{{appPackageSegment}}.feature.users

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import io.github.ydonghao.yarch.client.error.ApiError
import io.github.ydonghao.yarch.client.error.NetworkError
import io.github.ydonghao.{{appPackageSegment}}.core.network.api.UserDto
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * 状态下降 / 事件上行（UDF）：传输错误统一网络文案（client-shared 一-4），
 * 业务错误透出 message；取消（CancellationException）原生传导不捕获。
 */
@HiltViewModel
class UsersViewModel @Inject constructor(
    private val repository: UsersRepository
) : ViewModel() {

    data class UiState(
        val loading: Boolean = false,
        val users: List<UserDto> = emptyList(),
        val errorMessage: String? = null
    )

    private val _state = MutableStateFlow(UiState())
    val state: StateFlow<UiState> = _state.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        if (_state.value.loading) {
            return
        }
        viewModelScope.launch {
            _state.update { it.copy(loading = true, errorMessage = null) }
            try {
                val users = repository.users(1)
                _state.update { it.copy(loading = false, users = users) }
            } catch (error: ApiError) {
                _state.update { it.copy(loading = false, errorMessage = error.message) }
            } catch (_: NetworkError) {
                _state.update { it.copy(loading = false, errorMessage = "网络异常，请稍后重试") }
            }
        }
    }
}

File: clients/create/templates/ios16/Packages/Features/Users/Sources/Users/UsersViewModel.swift
import ContractKit
import Foundation

/**
 * MVVM ios16 扩展档（contract/clients/ios.md 二-2 降级表）：ObservableObject + @Published 同构实现。
 * 传输错误统一网络文案（client-shared 一-4），业务错误透出 message；取消原生传导不落错误态。
 */
@MainActor
public final class UsersViewModel: ObservableObject {
    public struct State: Sendable, Equatable {
        public var loading = false
        public var users: [User] = []
        public var errorMessage: String?

        public init() {}
    }

    @Published public var state = State()

    private let repository: any UsersRepository

    public init(repository: any UsersRepository) {
        self.repository = repository
    }

    public func refresh() async {
        guard !state.loading else { return }
        state.loading = true
        state.errorMessage = nil
        do {
            state.users = try await repository.users(page: 1)
        } catch let error as ApiError {
            state.errorMessage = error.message
        } catch is NetworkError {
            state.errorMessage = "网络异常，请稍后重试"
        } catch is CancellationError {
            // 取消不是错误（client-shared 一-4）：随 VM 生命周期静默收尾，不落错误态
            state.loading = false
            return
        } catch {
            state.errorMessage = "发生未知错误"
        }
        state.loading = false
    }
}
